app/Connect/Actions/SubmitGameTitleAction.php

```diff
@@ -0,0 +1,148 @@
+                return $this->accessDenied('You must be a developer to perform this action.');
+            }
+        }
+
+        if (mb_strlen($this->hash) != 32) {
+            return $this->invalidParameter('Hash must be 32 characters long.');
+        }
+
+        if (mb_strlen($this->gameTitle) < 2) {
+            return $this->invalidParameter('Title must be at least two characters long.');
+        }
+
+        if ($this->gameId) {
+            // trying to add hash to existing game
+            $game = Game::find($this->gameId);
+            if (!$game) {
+                return $this->gameNotFound();
+            }
+            if (!$game->system->active && !$this->user->hasRole(Role::GAME_HASH_MANAGER)) {
+                return $this->accessDenied('You do not have permission to add hashes to games for an inactive system.');
+            }
+        } else {
+            $system = System::find($this->systemId);
+            if (!$system) {
+                return $this->invalidParameter('Unknown system ID.');
+            }
+            if (!$system->active && !$this->user->hasRole(Role::GAME_HASH_MANAGER)) {
+                return $this->accessDenied('You do not have permission to add games to an inactive system.');
+            }
+
+            // title must be unique.
+            // If the title already exists, just try to add the hash to the existing game
+            $release = GameRelease::query()
+                ->where('title', $this->gameTitle)
+                ->with('game')
+                ->whereHas('game.system', function ($query) {
+                    $query->where('ID', $this->systemId);
+                })
+                ->first();
+            if ($release) {
+                $game = $release->game;
+            } else {
+                // no title match, it's a new game
+                $game = Game::create(['Title' => $this->gameTitle, 'ConsoleID' => $this->systemId]);
+
+                // Create the initial canonical title in game_releases.
+                $game->releases()->create([
+                    'title' => $this->gameTitle,
+                    'is_canonical_game_title' => true,
+                ]);
+
+                // create an empty GameAchievementSet and AchievementSet
+                (new UpsertGameCoreAchievementSetFromLegacyFlagsAction())->execute($game);
+            }
+        }
+
+        // associate the new hash if it doesn't already exist
+        if (!GameHash::where('game_id', $game->id)->where('md5', $this->hash)->exists()) {
+            // associate hash to game
+            $gameHash = new GameHash([
+                'game_id' => $game->id,
+                'user_id' => $this->user->id,
+                'md5' => $this->hash,
+                'compatibility' => GameHashCompatibility::Compatible,
+            ]);
+            if (!empty($this->hashDescription)) {
+                $gameHash->name = $this->hashDescription;
+            }
+            $gameHash->save();
```

Comment:
Very minor edge case: between checking if a hash exists and then creating it, another inbound request could create the same hash, leading to a duplicate key error.

This can probably be mitigated with `firstOrCreate`:
```php
$gameHash = GameHash::firstOrCreate(
  ['game_id' => $game->id, 'md5' => $this->hash],
  [
    'user_id' => $this->user->id,
    'compatibility' => GameHashCompatibility::Compatible,
  ]
);
```